### PR TITLE
[TESTING] PR to test and work out new CUDA versions for CI

### PR DESCRIPTION
**Pull request by mike-wendt**

### CODE REVIEW

**Comment by tijyojwad** [comment]
@mike-wendt questions about the new setup -

1. The CUDA 10.1 py 3.6 jobs seems to failing because of a numpy installation error. Any chance the python in the conda env has numpy already?
https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/gpuCI-private%2Fclara-genomics-analysis%2Fprb%2Fclara-genomics-analysis-cpu-build/detail/clara-genomics-analysis-cpu-build/869/pipeline

2. The CUDA 9.0 GPU jobs seems to be using CUDA 10.1 based on the logs for cmake. Could it be that the symlink fr /usr/local/cuda in that image is pointing to /usr/local/cuda-10.1 instead of /usr/local/cuda-9.0?
````
-- Using CUDA 10.1 from /usr/local/cuda
-- Using CUDA 10.1 from /usr/local/cuda
```

**Comment by raydouglass** [comment]
> 1. The CUDA 10.1 py 3.6 jobs seems to failing because of a numpy installation error. Any chance the python in the conda env has numpy already?
> https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/gpuCI-private%2Fclara-genomics-analysis%2Fprb%2Fclara-genomics-analysis-cpu-build/detail/clara-genomics-analysis-cpu-build/869/pipeline

Yes the conda environment has numpy already.

> 2. The CUDA 9.0 GPU jobs seems to be using CUDA 10.1 based on the logs for cmake. Could it be that the symlink fr /usr/local/cuda in that image is pointing to /usr/local/cuda-10.1 instead of /usr/local/cuda-9.0?

Are you sure? Can you point to the logs?
I see this one and has the right CUDA version in cmake: https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/gpuCI-private%2Fclara-genomics-analysis%2Fprb%2Fclara-genomics-analysis-cpu-build/detail/clara-genomics-analysis-cpu-build/867/pipeline

> 3. Are these new test targets only enabled for merges to master branch? Or all jobs for ClaraGenomicsAnalysis?

Triggering the CUDA 9.0 & 10.1 is manual right now. We want to iron out the builds before enabling it globally.

**Comment by tijyojwad** [comment]
@raydouglass 
> Are you sure? Can you point to the logs?

https://gpuci.gpuopenanalytics.com/blue/rest/organizations/jenkins/pipelines/gpuCI-private/pipelines/clara-genomics-analysis/pipelines/prb/pipelines/clara-genomics-analysis-gpu-build/runs/431/log/?start=0

If you look at the CMake logs, it says using CUDA 10.1

> Triggering the CUDA 9.0 & 10.1 is manual right now. We want to iron out the builds before enabling it globally.

Perfect, just wanted to make sure that out other merges won't be blocked till these issues are ironed out.

**Comment by tijyojwad** [comment]
@raydouglass @mike-wendt made changes to support CUDA 9.0. can you re-trigger this PR with the different CUDA 9.0 and CUDA 10.1 jobs? I realized that what we'll actually need is CUDA 9.0, CUDA 9.2 and CUDA 10.1 ideally

File: common/logging/include/claragenomics/logging/logging.hpp [hunk 1]
```diff
@@ -69,7 +71,13 @@
 #endif
 
 // MUST come after the defines of the logging level!
+#if !(CGA_CUDA_BEFORE_9_2)
```
Comment (ahehn-nv) [inline common/logging/include/claragenomics/logging/logging.hpp]: I would stick with `#ifdef CGA_CUDA_BEFORE_`, i.e. use `#ifdef` and not invert the bool, such that the special handling happens right after the `#ifdef`.

File: common/logging/include/claragenomics/logging/logging.hpp [hunk 2]
```diff
@@ -103,35 +111,55 @@ LoggingStatus SetHeader(bool logTime, bool logLocation);
...
+#else
 #define CGA_LOG_INFO(...) SPDLOG_INFO(__VA_ARGS__)
+#endif
 
 /// \ingroup logging
 /// \def CGA_LOG_WARN
 /// \brief Log at warning level
 ///
 /// parameters as per https://github.com/gabime/spdlog/blob/v1.x/README.md
+#if CGA_CUDA_BEFORE_9_2
```
Comment (ahehn-nv) [inline common/logging/include/claragenomics/logging/logging.hpp]: use `#ifdef`

File: common/logging/include/claragenomics/logging/logging.hpp [hunk 3]
```diff
@@ -103,35 +111,55 @@ LoggingStatus SetHeader(bool logTime, bool logLocation);
...
+#else
 #define CGA_LOG_WARN(...) SPDLOG_WARN(__VA_ARGS__)
+#endif
 
 /// \ingroup logging
 /// \def CGA_LOG_ERROR
 /// \brief Log at error level
 ///
 /// parameters as per https://github.com/gabime/spdlog/blob/v1.x/README.md
+#if CGA_CUDA_BEFORE_9_2
```
Comment (ahehn-nv) [inline common/logging/include/claragenomics/logging/logging.hpp]: use `#ifdef`

File: common/logging/include/claragenomics/logging/logging.hpp [hunk 4]
```diff
@@ -103,35 +111,55 @@ LoggingStatus SetHeader(bool logTime, bool logLocation);
...
+#else
 #define CGA_LOG_ERROR(...) SPDLOG_ERROR(__VA_ARGS__)
+#endif
 
 /// \ingroup logging
 /// \def CGA_LOG_CRITICAL
 /// \brief Log at fatal/critical error level (does NOT exit)
 ///
 /// parameters as per https://github.com/gabime/spdlog/blob/v1.x/README.md
+#if CGA_CUDA_BEFORE_9_2
```
Comment (ahehn-nv) [inline common/logging/include/claragenomics/logging/logging.hpp]: use `#ifdef`

File: common/utils/include/claragenomics/utils/cudaversions.hpp [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+#if (__CUDACC_VER_MAJOR__ == 9 && __CUDACC_VER_MINOR__ < 2)
+#define CGA_CUDA_BEFORE_9_2 1
+#else
+#define CGA_CUDA_BEFORE_9_2 0
+#endif
+
+/// \def CGA_CONSTEXPR
+/// \brief C++ constexpr for device code - falls back to const for CUDA 10.0 and earlier
+#if CGA_CUDA_BEFORE_10_1
```
Comment (ahehn-nv) [inline common/utils/include/claragenomics/utils/cudaversions.hpp]: use `#ifdef`

Comment (ahehn-nv) [reply common/utils/include/claragenomics/utils/cudaversions.hpp]: (i.e. correct my code ;) )

**Comment by raydouglass** [comment]
rerun tests

**Comment by raydouglass** [comment]
rerun tests

**Comment by raydouglass** [comment]
rerun tests

**Comment by tijyojwad** [comment]
New set of tests are passing. Failing tests are old status checks that can't be removed.

**Pull request merged**